langchain/callbacks/mlflow_callback.py

```diff
@@ -0,0 +1,691 @@
+        """To finish the run."""
+        self.mlfc.set_terminated(self.run.info.run_id)
+
+    def metric(self, key: str, value: float):
+        """To log metric to mlflow server."""
+        # metric value only of type float or int
+        if not isinstance(value, float) and not isinstance(value, int):
+            return
+        self.mlfc.log_metric(self.run.info.run_id, key, value)
+
+    def json(self, data: Any):
+        """To log all metrics in the input dict."""
+        for k, v in data.items():
+            self.metric(k, v)
+
+    def jsonf(self, data: Any, filename: str):
+        """To log the input data as json file artifact."""
+        self.mlfc.log_dict(self.run.info.run_id, data, f"{filename}.json")
+
+    def table(self, name: str, df):
+        """To log the input dataframe as a html table"""
+        self.html(df.to_html(), f"table_{name}")
+
+    def html(self, html: str, filename: str):
+        """To log the input html string as html file artifact."""
+        self.mlfc.log_text(self.run.info.run_id, html, f"{filename}.html")
+
+    def text(self, text: str, filename: str):
+        """To log the input text as text file artifact."""
+        self.mlfc.log_text(self.run.info.run_id, text, f"{filename}.txt")
+
+    def artifact(self, path: str):
+        """To upload the file from given path as artifact."""
+        self.mlfc.log_artifact(self.run.info.run_id, path)
+
+
+class MlflowCallbackHandler(BaseCallbackHandler):
+    """Callback Handler that logs metrics and artifacts to mlflow server.
+
+    Parameters:
+        name (str): Name of the run.
+        experiment (str): Name of the experiment.
+        tags (str): Tags to be attached for the run.
+        tracking_uri (str): MLflow tracking server uri.
+
+    This handler will utilize the associated callback method called and formats
+    the input of each callback function with metadata regarding the state of LLM run,
+    and adds the response to the list of records for both the {method}_records and
+    action. It then logs the response to mlflow server.
+    """
+
+    def __init__(
+        self,
+        name: Optional[str] = "langchainrun-%",
+        experiment: Optional[str] = "langchain",
+        tags: Optional[Sequence] = [],
+        tracking_uri: Optional[str] = None,
+    ) -> None:
+        """Initialize callback handler."""
+        import_pandas()
+        import_textstat()
+        spacy = import_spacy()
+        super().__init__()
+
+        self.name = name
+        self.experiment = experiment
+        self.tags = tags
+        self.tracking_uri = tracking_uri
+
+        self.temp_dir = tempfile.TemporaryDirectory()
+
+        self.mlflg = MlflowLogger(
+            tracking_uri=self.tracking_uri,
+            experiment_name=self.experiment,
+            run_name=self.name,
+            run_tags=self.tags,
+        )
+
+        self.action_records: list = []
+        self.nlp = spacy.load("en_core_web_sm")
+
+        self.metrics = {

```

Comment:
> Hi @mahmedk, excited to see your efforts here! I'm from the MLflow maintainers team and we are actively working on adding new MLflow APIs to support recording text inputs, outputs, and LLM prompts, and it would be released with MLflow soon. With the new APIs, this PR could be largely simplified. I'm also working on the LangChain integration with MLflow and I'm happy to collaborate on this topic.

Hi @liangz1 , Thanks. Glad to have you looking at this PR. Super excited to hear that there will be native support for tracking prompts in MLFlow. I will be more than happy to collaborate.

Reply:
> Dumb question: Could you help me understand in which scenario would the user find the metrics helpful, and why would the user want to log these metrics?

These are some basic counters which records the actions taken by langchain for a text generating flow setup by user. User could use these for debugging, analysis ,  check the state of langchain resources like agents, chain & llms such as if they were called and if their start/end calls match.